### PR TITLE
feat(docs): add migrating to React Hooks Blog Post

### PR DESCRIPTION
**Pull request by dslemay**
## Description

Adds a blog post discussing some of the benefits of converting some components to use React Hooks. Also touches on structuring React Hooks in a Gatsby project and some things to be aware of, including the `useEffect` hook and Gatsby's build process.

## Related Issues

Related to #11513

### CODE REVIEW

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+![Custom React Hook for Slideshow functionality](./images/custom-react-hook.jpg)
+
+The benefits of this structural change go beyond aesthetic cleanup. One major benefit is the cleanup and re-running of useEffect when specific properties change. This eliminates the need of storing the timeout id as a static property. It also eliminates imperative prop comparisons in `componentDidUpdate`. Instead, the function declares the data that will trigger a re-run of this side effect. If `currIndex` or `isPlaying` change, the effect will re-run. First the function will run the cleanup function to clear the timeout, and then will run the effect again. This compounds when a component requires many side effects.
```
Comment (DSchau) [inline docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: This is a great point! I do think that the change detection (and optimization) in hooks seems to be much improved. Nicely said!

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+Hooks bring more options on when to convert a functional component to a class based component if at all. Custom Hooks work well for abstracting complex logic. These provide the same composable logic as render props without the added component in the tree or syntax. Hooks improve the ability to share logic across components. You can find the code conversion to Hooks discussed in this post at the [related PR](https://github.com/dslemay/Portfolio-Site/pull/10). Hooks may not be the death of class based components, but they do encourage critical thought before reaching for a class component.
```
Comment (DSchau) [inline docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: This is a bit wordy. One thing I've found super helpful is to break up my sentences into a list structure when concluding, e.g

```
Quite a lot of benefits of hooks are now apparent. Specifically, in this discussion, the following benefits of hooks are now clear:

- benefit a
- benefit b
- benefit c

From this basis, the value of hooks are clear and obvious. I can't wait to see what you build with hooks, and I believe they constitute a meaningful shift in the way we do React development.
```

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+author: Daniel Lemay
+tags:
+  - react
+  - react hooks
+image: "./images/hooks-diff.jpg"
+showImageInArticle: true
+canonicalLink: "https://dslemay.com/blog/2019/02/04/pragmatic-lessons-from-converting-to-react-hooks"
+---
+
+Last week I decided to install the React 16.8 alpha on a branch and experiment with React Hooks in preparation for their release on February 4, 2018. The site utilized a [render prop](https://reactjs.org/docs/render-props.html) based Slideshow component in several places as well as a handful of other class based components. Through this process, I was able to consolidate the application code and eliminate all class based components from the site's code base.
```
[the remark this answers is not in the document]
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: Good call. Adding that clarification now.

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+The custom `useSlideshow` hook utilizes two different hooks to replace the functionality of the render props component: `useState` and `useEffect`. The current index and playing states are both set with their own calls to useState. The `useEffect` hook checks if the isPlaying state is true and then sets the timeout to advance the slide to the next index. It resets to the first slide after it reaches the last index. The hook clears the timeout when the current index or isPlaying state changes. The hook includes a function to update the the slide. The necessary state and functions are return in an object.
+
+![Custom React Hook for Slideshow functionality](./images/custom-react-hook.jpg)
```
[the remark this answers is not in the document]
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: Sounds good. Making changes in next commit.

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md (hunk 2 above)
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: Updated in latest commit.

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+---
+
+Last week I decided to install the React 16.8 alpha on a branch and experiment with React Hooks in preparation for their release on February 4, 2018. The site utilized a [render prop](https://reactjs.org/docs/render-props.html) based Slideshow component in several places as well as a handful of other class based components. Through this process, I was able to consolidate the application code and eliminate all class based components from the site's code base.
+
+## Converting to Hooks and lessons learned
+
+The [Hooks documentation](https://reactjs.org/docs/hooks-intro.html) is well presented and an excellent resource to getting started with hooks. I previously championed the render props pattern for reusable logic and composability. However, the extra syntax often comes with tradeoffs in clarity. These concerns include: "wrapper hell" and more mental overhead to parse nested JSX structure within the component. Adding a second render prop component or ternary operator further compounds these concerns. These additional wrappers also display in the React devTools and can get out of hand. The possibility of cleaning up this syntax and providing clearer code is alluring.
```
[the remark this answers is not in the document]
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: Added a quick screenshot for visual reference.

[the remark this answers is not in the document]
Comment (DSchau) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: I think actual source code that creates that devtools structure is perhaps clearer?

**Comment by dslemay** [comment]
@DSchau I believe that the comments have all been addressed. Let me know if anything else needs further massaging or clarification.

A quick side note is that I was waiting to publish the version on my site until Hooks was actually released to stable. I can adapt that as well if it is desired to have this go out sooner. Just want to make sure that the canonicalUrl actually exists.

**Comment by dslemay** [comment]
Sounds good. I have been keeping a tab on the npm page. I have everything queued up in Contentful on my site's side as well. Just need to update the branch to stable, merge, and publish.

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
+---
+title: "Pragmatic Lessons from Converting to React Hooks"
+date: 2019-02-04
```
[the remark this answers is not in the document]
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: That works for me. `canonicalLink` will remain for today. Change forthcoming.

File: docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+## Structuring Hooks in Gatsby
+
+Currently I am structuring all my custom hooks in a top level folder so that other components can import them from a central location. Pulling the React dependencies out of Gatsby in version 2 allows for using Hooks immediately. To begin using Hooks today, update React and React-DOM to 16.8.0. There are considerations to take with the `useEffect Hook`. If it references the window object, you need to check that window is defined to avoid Gatsby build errors. These effects would normally live in `componentDidMount` where the component hydrates in the DOM. Hooks are called in the build process. The Gatsby docs have great resources for [debugging HTML builds](docs/debugging-html-builds/) if you encounter this issue.
+
+```javascript
+function useMediaQuery() {
```
[the remark this answers is not in the document]
Comment (DSchau) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: (Note: I swapped this a bit. I like to use min-width queries rather than max-width queries!)

[the remark this answers is not in the document]
Comment (dslemay) [reply docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md]: Definitely. This example was more simple since it fit the use case for this particular project.

One thing I have thought of doing on another project which has more complex needs is to have the useMediaQuery function accept an object where some common queries would check if the corresponding options key was set to true. If so it would add the corresponding listener. If false, it would not add the listener. There could potentially be optional keys for adding custom queries as well.

**Comment by dslemay** [comment]
Given that the post refers to the release date it probably should not be merged until that is confirmed since it may not be released today after all.

**Comment by DSchau** [comment]
@dslemay fair-ish, but I scheduled it for Thursday which seems pretty reasonable

**Comment by dslemay** [comment]
Definitely. My comment was more in reference to the first sentence of the post: "Last week I decided to install the React 16.8 alpha on a branch and experiment with React Hooks in preparation for their release on February 4, 2018." If release date ends up not being today this should either be removed or updated I would think.
